src/features/sourceGeneratedDocumentProvider.ts

```diff
@@ -0,0 +1,127 @@
+
+import * as serverUtils from '../omnisharp/utils';
+import { Event, EventEmitter, TextDocument, TextDocumentContentProvider, TextEditor, Uri, window, workspace } from 'vscode';
+import { IDisposable } from '../Disposable';
+import { SourceGeneratedFileInfo, SourceGeneratedFileResponse, UpdateType } from '../omnisharp/protocol';
+import { OmniSharpServer } from '../omnisharp/server';
+
+export default class SourceGeneratedDocumentProvider implements TextDocumentContentProvider, IDisposable {
+    readonly scheme = "omnisharp-source-generated";
+    private _registration: IDisposable;
+    private _documents: Map<SourceGeneratedFileInfo, SourceGeneratedFileResponse>;
+    private _uriToDocumentInfo: Map<string, SourceGeneratedFileInfo>;
+    private _documentClosedSubscription: IDisposable;
+    private _visibleTextEditorsChangedSubscription: IDisposable;
+    private _onDidChangeEmitter: EventEmitter<Uri>;
+    public onDidChange: Event<Uri>;
+
+    constructor(private server: OmniSharpServer) {
+        this._documents = new Map<SourceGeneratedFileInfo, SourceGeneratedFileResponse>();
+        this._uriToDocumentInfo = new Map<string, SourceGeneratedFileInfo>();
+        this._documentClosedSubscription = workspace.onDidCloseTextDocument(this.onTextDocumentClosed, this);
+        this._visibleTextEditorsChangedSubscription = window.onDidChangeVisibleTextEditors(this.onVisibleTextEditorsChanged, this);
+        this._onDidChangeEmitter = new EventEmitter<Uri>();
+        this.onDidChange = this._onDidChangeEmitter.event;
+    }
+
+    private async onTextDocumentClosed(document: TextDocument) {
+        const uriString = document.uri.toString();
+        if (this._uriToDocumentInfo.has(uriString)) {
+            const info = this._uriToDocumentInfo.get(uriString);
+            this._documents.delete(info);
+            this._uriToDocumentInfo.delete(uriString);
+            await serverUtils.sourceGeneratedFileClosed(this.server, info);
+        }
+    }
+
+    private async onVisibleTextEditorsChanged(editors?: TextEditor[]) {
+        for (const editor of editors) {
+            const documentUri = editor.document.uri;
+            const uriString = documentUri.toString();
+            if (this._uriToDocumentInfo.has(uriString)) {
+                try {
+                    const existingInfo = this._uriToDocumentInfo.get(uriString);
+                    const existingResponse = this._documents.get(existingInfo);
+                    const update = await serverUtils.getUpdatedSourceGeneratedFile(this.server, existingInfo);
+                    if (!update) {
+                        continue;
+                    }
+
+                    switch (update.UpdateType) {
+                        case UpdateType.Deleted:
+                            this._documents.set(existingInfo, { Source: "Document is no longer being generated.", SourceName: existingResponse.SourceName });
```

Comment:
should it not be evicted from `_uriToDocumentInfo` at this point too?

Reply:
No. The document is still open in the editor, and we need to be able to display this text when you switch to it. We'll evict the document when the tab is closed.